beetsplug/substitute.py

```diff
@@ -46,11 +48,32 @@ def __init__(self):
         Get the configuration, register template function and create list of
         substitute rules.
         """
-        super().__init__()
+        super(Substitute, self).__init__()
         self.substitute_rules = []
         self.template_funcs["substitute"] = self.tmpl_substitute
 
-        for key, view in self.config.items():
-            value = view.as_str()
-            pattern = re.compile(key.lower())
+        if isinstance(self.config.get(), dict):
+            self._log.warning(
+                "Unordered configuration is deprecated, as it leads to"
+                + " unpredictable behaviour on overlapping rules.\n"
```

Comment:
I'm slightly confused about this issue. Using the old configuration, does the issue have to do with a configuration like this?

```yaml
substitute:
  a: b
  a: c
```

If so, I think it should be pretty much predictable: the last duplicate key wins, since it's equivalent to `{"a": "b", "a": "c"}` in Python.

Reply:
This ties in to the above question, that we only apply the first matching rule. Because our keys are regex patterns, multiple *different* keys can match the same input. With the below configuration, what do you expect `a` to be substituted with?
```yaml
substitute:
  a: b
  (a|x): c
```

I initially believed it would result in `b`, since that's the first defined rule, but in testing I found it to be `c`.

I now believe it is because the keys end up in alphabetical order (or at least some other order), which makes it hard to predict or control which rule should be applied. By turning it into a list, we preserve the order of rules as defined by the user. Behaviour then becomes simple to define: the first matching rule is applied.

Reply:
How did you test this? And most importantly, which Python version are you using?

`config["substitute"]` returns an `OrderedDict`

```python
>>> from beets import config

>>> list(config["substitute"].items())
Out[11]: [('a', <Subview: substitute.a>), ('(a|x)', <Subview: substitute.(a|x)>)]

>>> config["substitute"].flatten()
Out[12]: OrderedDict([('a', 'b'), ('(a|x)', 'c')])
```

So I'd expect the plugin to apply the replacements in the sequence they are defined. And it seems like it's working as expected:

```python
>>> from beetsplug.substitute import Substitute

>>> s = Substitute()

>>> s.substitute_rules
Out[7]: [(re.compile(r'a', re.UNICODE), 'b'), (re.compile(r'(a|x)', re.UNICODE), 'c')]

>>> s.tmpl_substitute("a")
Out[8]: 'b'
```

Reply:
I encountered it in one of the test cases I had written, on my own Python 3.12 setup and in the [Github Actions tests](https://github.com/beetbox/beets/actions/runs/11093855139/job/30820644945#step:8:121) in this PR. 

I can however replicate your findings on my machine, and it seems the order is in fact preserved when loaded from the yaml configuration. Perhaps this is then only an error in the testing setup.

My attempts to fix the test case however have not been successful. Even declaring the configuration as an OrderedDict still results in a jumbled up configuration. 

```py
def test_break_on_first_match(self):
    self.run_substitute(
        OrderedDict([
            ("a", "x"),
            ("[ab]", "y"),
        ]),
        [
            ("a", "x"),
            ("b", "y"),
        ],
    )
```

This test still fails the first case, as `a` becomes substituted with `y`.

Reply:
Ahh, I see. Have tested it and found that the issue is with how the configuration is applied in the tests:

```python
    @contextmanager
    def configure_plugin(self, config: list[Any] | dict[str, Any]):
        if isinstance(config, list):
            beets.config[self.plugin] = config
        else:
            for key, value in config.items():
                beets.config[self.plugin][key] = value
        self.load_plugins(self.plugin)

        yield

        self.unload_plugins()
```

Apologies - I've written this context manager myself and back then I did not know the inner working of `beets.config` yet.

Try replacing this function in `beets.test.helper` with

```python
    @contextmanager
    def configure_plugin(self, config: Any):
        beets.config[self.plugin].set(config)
        self.load_plugins(self.plugin)

        yield

        self.unload_plugins()
```

Reply:
> Apologies - I've written this context manager myself and back then I did not know the inner working of `beets.config` yet.

Haha, no worries! Thank you very much for figuring out where it goes wrong :)

This does simplify this PR quite a lot then, if I can skip the change of configuration format. Should I make your above change as part of this PR, or should it be a separate change?